src/library/tools/pluginServer/index.ts

```diff
@@ -1,92 +1,75 @@
-  .then((activePlugins) => {
-    // The type of `POST /verifyProof` requests' body.
-    interface VerifyProofData {
-      plugin: string;
-      publicInputArgs: unknown;
-      proof: JsonProof;
-    }
-
-    // Use the appropriate plugin to verify the proof and return the output.
-    async function verifyProof(data: VerifyProofData): Promise<unknown> {
-      const pluginName = data.plugin;
-      console.info(`verifying proof using plugin ${pluginName}`);
-      const pluginInstance = activePlugins[pluginName];
-      if (!pluginInstance) throw `plugin ${pluginName} not found`;
-      // Step 1: check that the proof is valid against the given verification key.
-      const proofValid = await verify(
-        data.proof,
-        pluginInstance.verificationKey
-      );
-      if (!proofValid) throw `invalid proof`;
-      // Step 2: use the plugin to extract the output. The plugin is also responsible
-      // for checking the legitimacy of the public inputs.
-      const typedPublicInputArgs = pluginInstance.publicInputArgsSchema.parse(
-        data.publicInputArgs
-      );
-      const output = await pluginInstance.verifyAndGetOutput(
-        typedPublicInputArgs,
-        data.proof
-      );
-      return output;
-    }
-
-    const app = express().use(bodyParser.json());
-
-    // Register all custom routes of active plugins under `/plugins/${pluginName}`.
-    Object.entries(activePlugins).map(([name, plugin]) =>
-      Object.entries(plugin.customRoutes).map(([path, handler]) =>
-        app.use(`/plugins/${name}/${path}`, handler)
+const main = pipe(
+  TE.Do,
+  TE.bind('cfg', () => readConfigurationFallback),
+  TE.let('pcProvider', () => new InMemoryProofCacheProvider()),
+  TE.tap(() => TE.fromIO(() => console.info('initializing plugins'))),
+  TE.bind('activePlugins', ({ cfg, pcProvider }) =>
+    initializePlugins(cfg, pcProvider)
+  ),
+  TE.bind('app', () => TE.fromIO(express)),
+  TE.tap(({ app, activePlugins, cfg }) =>
+    pipe(
+      TE.fromIO(() => {
+        app.use(bodyParser.json());
+        TE.fromIO(() => console.info('installing custom routes for plugins'));
+      }),
+      TE.chain(() => installCustomRoutes(activePlugins)(app)),
+      TE.chain(() =>
+        TE.fromIO(() => {
+          app
+            .post(
+              '/verifyProof',
+              (req: Request, res: Response): Promise<void> => {
+                const body = req.body as VerifyProofData;
+                return pipe(
+                  verifyProof(activePlugins)(
+                    body.proof,
+                    body.publicInputArgs,
+                    body.plugin
+                  ),
+                  T.chain(
+                    E.match(
+                      (error: string) =>
+                        T.fromIO(() => res.status(400).json({ error })),
+                      (r) => T.fromIO(() => res.status(200).json(r))
+                    )
+                  ),
+                  T.map(() => {})
```

Comment:
What's this?

Reply:
Basically, if you write out the type of the resulting lambda it would be `(_: TaskEither<string, Express>): TaskEither<string, void>`.